src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs

```diff
@@ -0,0 +1,112 @@
+using Diagnostics.DataProviders.TokenService;
+using System.Diagnostics;
+using Microsoft.Azure.OperationalInsights.Models;
+using Diagnostics.Logger;
+
+namespace Diagnostics.DataProviders
+{
+    class K8SELogAnalyticsClient : ILogAnalyticsClient
+    {
+        private Microsoft.Rest.ServiceClientCredentials creds;
+        private string workspaceId;
+        private static OperationalInsightsDataClient client;
+        private string _requestId;
+
+        public K8SELogAnalyticsClient(string RequestId)
+        {
+            creds = K8SELogAnalyticsTokenService.Instance.getCreds();
+            workspaceId = K8SELogAnalyticsTokenService.Instance.getWorkspaceId();
+            client = new OperationalInsightsDataClient(creds);
+            client.WorkspaceId = workspaceId;
+            _requestId = RequestId;
+        }
+
+        public async Task<DataTable> ExecuteQueryAsync(string query)
+        {
+            var timeTakenStopWatch = new Stopwatch();
+            QueryResults queryResults;
+            Exception dataProviderException = null;
+            //check query
+
+            try
+            {
+                timeTakenStopWatch.Start();
+                queryResults = await client.QueryAsync(query);
+            }
+            catch (Exception ex)
+            {
+                dataProviderException = ex;
+                throw;
+            }
+            finally
+            {
+                timeTakenStopWatch.Stop();
+                var StopTime = DateTime.Now;
+                var StartTime = StopTime.AddMilliseconds(-timeTakenStopWatch.ElapsedMilliseconds);
+
+                if (dataProviderException != null)
+                {
+                    DiagnosticsETWProvider.Instance.LogDataProviderException(_requestId, 
+                        "K8SELogAnalyticsDataProvider", StartTime.ToString("HH:mm:ss.fff"),
+                        StopTime.ToString("HH:mm:ss.fff"), timeTakenStopWatch.ElapsedMilliseconds,
+                        dataProviderException.GetType().ToString(), dataProviderException.ToString());
+                }
+                else
+                {
+                    DiagnosticsETWProvider.Instance.LogDataProviderOperationSummary(_requestId,
+                        "K8SELogAnalyticsDataProvider", StartTime.ToString("HH:mm:ss.fff"),
+                        StopTime.ToString("HH:mm:ss.fff"), timeTakenStopWatch.ElapsedMilliseconds);
+                }
+
+            }
+
+            if (queryResults == null)
+            {
+                queryResults = new QueryResults();
+            }
+            var dataTable = ResultAsDataTable(queryResults);
+            return dataTable;
+        }
+
+        //convert results to DataTable
+        private DataTable ResultAsDataTable(QueryResults results)
+        {
+            DataTable dataTable = new DataTable("results");
+            dataTable.Clear();
+
+            int cursor = 0;
+            foreach (var i in results.Tables[0].Columns)
```

Comment:
why do you need cursor variable here. You can simply use i variable in foreach to access the columns[cursor].
Regardless, explore System.Linq in c# to achieve these things in a much more smaller and compact code.
https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.selectmany?view=net-5.0

Reply:
Also, since the structures of Data Table and results.Table[0] seems very similar, see if you can simply typecast the object (if that works)